deps/status/src/checkLockfilesUpToDate.ts

```diff
@@ -0,0 +1,380 @@
+| 'hooks'
+| 'peersSuffixMaxLength'
+| 'rootProjectManifest'
+| 'rootProjectManifestDir'
+| 'sharedWorkspaceLockfile'
+| 'workspaceDir'
+>
+
+export async function checkLockfilesUpToDate (opts: CheckLockfilesUpToDateOptions): Promise<void> {
+  const {
+    allProjects,
+    autoInstallPeers,
+    catalogs,
+    excludeLinksFromLockfile,
+    linkWorkspacePackages,
+    rootProjectManifest,
+    rootProjectManifestDir,
+    sharedWorkspaceLockfile,
+    workspaceDir,
+  } = opts
+
+  const rootManifestOptions = rootProjectManifest && rootProjectManifestDir
+    ? getOptionsFromRootManifest(rootProjectManifestDir, rootProjectManifest)
+    : undefined
+
+  if (allProjects && workspaceDir) {
+    const packagesList = await loadPackagesList(workspaceDir)
+    if (!packagesList) {
+      throw new PnpmError('RUN_CHECK_DEPS_NO_CACHE', 'Cannot check whether dependencies are outdated', {
+        hint: 'Run `pnpm install` to create the cache',
+      })
+    }
+
+    if (!equals(
+      filter(value => value != null, packagesList.catalogs ?? {}),
+      filter(value => value != null, catalogs ?? {})
+    )) {
+      throw new PnpmError('RUN_CHECK_DEPS_OUTDATED', 'Catalogs cache outdated', {
+        hint: 'Run `pnpm install` to update the catalogs cache',
+      })
+    }
+
+    const currentProjectRootDirs = allProjects.map(project => project.rootDir).sort()
+    if (!equals(packagesList.projectRootDirs, currentProjectRootDirs)) {
+      throw new PnpmError('RUN_CHECK_DEPS_WORKSPACE_STRUCTURE_CHANGED', 'The workspace structure has changed since last install', {
+        hint: 'Run `pnpm install` to update the workspace structure and dependencies tree',
+      })
+    }
+
+    const allManifestStats = await Promise.all(allProjects.map(async project => {
+      const manifestStats = await statManifestFile(project.rootDir)
+      if (!manifestStats) {
+        // this error should not happen
+        throw new Error(`Cannot find one of ${MANIFEST_BASE_NAMES.join(', ')} in ${project.rootDir}`)
+      }
+      return { project, manifestStats }
+    }))
+
+    const modifiedProjects = allManifestStats.filter(
+      ({ manifestStats }) =>
+        manifestStats.mtime.valueOf() > packagesList.lastValidatedTimestamp
+    )
+
+    if (modifiedProjects.length === 0) {
+      logger.debug({ msg: 'No manifest files are modified after the last validation. Exiting check.' })
+      return
+    }
+
+    logger.debug({ msg: 'Some manifest files are modified after the last validation. Continuing check.' })
+
+    let readWantedLockfileAndDir: (projectDir: string) => Promise<{
+      wantedLockfile: Lockfile
+      wantedLockfileDir: string
+    }>
+    if (sharedWorkspaceLockfile) {
+      const wantedLockfileStats = await statIfExists(path.join(workspaceDir, WANTED_LOCKFILE))
+      if (!wantedLockfileStats) return throwLockfileNotFound(workspaceDir)
+
+      const wantedLockfilePromise = readWantedLockfile(workspaceDir, { ignoreIncompatible: false })
```

Comment:
Size does not imply equality. For example: `foo: 1.0.0` and `foo: 2.0.0` have the exact same size but different values. Of course, if the sizes are not equal, then the lockfiles are likely not equal, but that would increase code complexity just to optimize for an error path. Is this a worthwhile trade-off?

Reply:
It won't be an error path as as we discussed, it will run install in that case.

But it looks like we will need to read the lockfile anyway to support filtered install.